microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/ls/commons/snippets/SnippetRegistry.java

```diff
@@ -228,86 +311,94 @@ private static String getInsertText(Snippet snippet, boolean replace, String lin
 				if (i > 0) {
 					text.append(lineDelimiter);
 				}
-				if (replace) {
-					bodyLine = replace(bodyLine);
-				}
+				bodyLine = merge(bodyLine, model, replace);
 				text.append(bodyLine);
 				i++;
 			}
 		}
 		return text.toString();
 	}
 
-	private static String replace(String line) {
-		return replace(line, 0, null);
+	private static String merge(String line, Map<String, String> model, boolean replace) {
+		return replace(line, 0, model, replace, null);
 	}
 
-	private static String replace(String line, int offset, StringBuilder newLine) {
-		int startExpr = line.indexOf("${", offset);
-		if (startExpr == -1) {
+	private static String replace(String line, int offset, Map<String, String> model, boolean replace,
+			StringBuilder newLine) {
+		int dollarIndex = line.indexOf("$", offset);
+		if (dollarIndex == -1 || dollarIndex == line.length() - 1) {
 			if (newLine == null) {
 				return line;
 			}
 			newLine.append(line.substring(offset, line.length()));
 			return newLine.toString();
 		}
-		int endExpr = line.indexOf("}", startExpr);
-		if (endExpr == -1) {
-			// Should never occur
-			return line;
-		}
 		if (newLine == null) {
 			newLine = new StringBuilder();
 		}
-		newLine.append(line.substring(offset, startExpr));
-		// Parameter
-		int startParam = startExpr + 2;
-		int endParam = endExpr;
-		boolean startsWithNumber = true;
-		for (int i = startParam; i < endParam; i++) {
-			char ch = line.charAt(i);
-			if (Character.isDigit(ch)) {
-				startsWithNumber = true;
-			} else if (ch == ':') {
-				if (startsWithNumber) {
-					startParam = i + 1;
-				}
-				break;
-			} else if (ch == '|') {
-				if (startsWithNumber) {
-					startParam = i + 1;
-					int index = line.indexOf(',', startExpr);
-					if (index != -1) {
-						endParam = index;
+		char next = line.charAt(dollarIndex + 1);
+		if (Character.isDigit(next)) {
+			if (replace) {
+				newLine.append(line.substring(offset, dollarIndex));
+			}
+			int lastDigitOffset = dollarIndex + 1;
+			while (line.length() < lastDigitOffset && Character.isDigit(line.charAt(lastDigitOffset))) {
+				lastDigitOffset++;
+			}
+			if (!replace) {
+				newLine.append(line.substring(offset, lastDigitOffset));
+			}
+			return replace(line, lastDigitOffset, model, replace, newLine);
+		} else if (next == '{') {
+			int startExpr = dollarIndex;
+			int endExpr = line.indexOf("}", startExpr);
+			if (endExpr == -1) {
+				// Should never occur
+				return line;
+			}
+			newLine.append(line.substring(offset, startExpr));
+			// Parameter
+			int startParam = startExpr + 2;
+			int endParam = endExpr;
+			boolean startsWithNumber = true;

```

Comment:
`startsWithNumber` is initialized with `true` only ever set to `true`

Reply:
Good catch! I removed it.